pkg/cri/sbserver/restart.go

```diff
@@ -82,6 +82,28 @@ func (c *criService) recover(ctx context.Context) error {
 		return err
 	}
 
+	// Recover sandboxes in the new SandboxStore
+	storedSandboxes, err := c.client.SandboxStore().List(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list sandboxes from API: %w", err)
+	}
+	for _, sbx := range storedSandboxes {
+		if _, err := c.sandboxStore.Get(sbx.ID); err == nil {
+			continue
+		}
+		metadata := sandboxstore.Metadata{}
+		err := sbx.GetExtension("metadata", &metadata)
+		if err != nil {
+			return fmt.Errorf("failed to get metadata for stored sandbox %q: %w", sbx.ID, err)
+		}
+		sb := sandboxstore.NewSandbox(metadata, sandboxstore.Status{State: sandboxstore.StateUnknown})
```

Comment:
Why initialize as sandboxstore.StateUnknown? What if the stored sandbox are in ready state?

Reply:
The existing container-based restore code is still above this and should capture the cases where the state is known (same logic as the `server` implementation).  This codepath covers the other cases that exist in `sbserver` now, since the actual container creation happens in a different part of the logic.  When the container doesn't exist, either we're restoring after an incomplete start or an incomplete stop, but we don't know which.  Setting to `StateUnknown` allows for a transition to `StateNotReady` and then Removal (if I understand the [state diagram](https://github.com/containerd/containerd/blob/ac5e72442b513ab3f4dfb813785be1dac2e12062/pkg/cri/store/sandbox/status.go#L27) properly).

The equivalent container-based restore code prefers [`StateNotReady`](https://github.com/containerd/containerd/blob/ac5e72442b513ab3f4dfb813785be1dac2e12062/pkg/cri/server/restart.go#L393-L395) for this scenario.  I can change the code here to use `StateNotReady` if you think that's better.

Reply:
Ah I get it. We skip the `sbx` if it is already in `sandboxStore`.

For `StateUnknown` vs `StateNotReady`, let's make it consistent in the follow-up. The state diagram does state it should be `StateUnknown` but the `pkg/cri/server/restart.go` chooses `StateNotReady`.

Captured in https://github.com/containerd/containerd/issues/7441